portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,35 +14,85 @@
 
 package com.google.sps.servlets;
 
+import java.util.*;
 import com.google.gson.Gson;
-import com.google.common.collect.ImmutableList;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that returns some example content. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private static final ImmutableList<String> FRUIT_FACTS =
-    ImmutableList.of(
-      "A strawberry is not an actual berry", 
-      "Kiwis contain more Vitamin C than oranges", 
-      "The pineapple is actually a berry");
+  private HashMap<String, String> OLD_FACTS = new HashMap<String, String>();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String fruitFactsJson = convertToJsonUsingGson(FRUIT_FACTS);
+    String oldFactsJson = convertToJsonUsingGson(OLD_FACTS);
 
     // Send the JSON as the response
     response.setContentType("application/json;");
-    response.getWriter().println(fruitFactsJson);
+    response.getWriter().println(oldFactsJson);
   }
 
-  private String convertToJsonUsingGson(ImmutableList<String> fruitFacts) {
+  private String convertToJsonUsingGson(HashMap<String, String> oldFacts) {
     Gson gson = new Gson();
-    String json = gson.toJson(fruitFacts);
+    String json = gson.toJson(oldFacts);
     return json;
   }
-}
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    String userFirstName = getParameter(request, "first-name", "");
+    String userLastName = getParameter(request, "last-name", "");
+    String userFunFact = getParameter(request, "user-fun-fact", "");
+
+    // Define key and value pair to add to OLD_FACTS
+    String factKey;
+    if (userFirstName.isEmpty() && userLastName.isEmpty()) {
+      factKey = "Anonymous";
+    }
+    else if (userFirstName.isEmpty()) {
+      factKey = userLastName;
+    }
+    else if (userLastName.isEmpty()) {
+      factKey = userFirstName;
+    }
+    else {
+      factKey = userFirstName + " " + userLastName;
+    }
+    String factValue = "";
+    if (!(userFunFact.isEmpty())) {
+      if (OLD_FACTS.containsKey(factKey)) {
+        factValue = OLD_FACTS.get(factKey) + ", " + userFunFact;
+        OLD_FACTS.remove(factKey);
+      }
+      else {
+        factValue = userFunFact;
+      }
+    }
+    
+    // Store user's input to access later
+    if (!(userFunFact.isEmpty())) {
+      OLD_FACTS.put(factKey, factValue);
+    }
+
+    // Redirect back to the HTML page.
+    response.sendRedirect("/index.html");
+    response.getWriter().println(factKey + ": " + factValue);
```

Comment:
Let's restructure the rest of this logic a bit.

First, it seems that even if factValue is empty (which probably means userFunFact is empty), we still add it to the response. Is this what we want? It might be better to just end early if userFunFact is empty (assuming this is the logic you want). This can be done via a quick check and early return. Otherwise, it might make sense to check if there's an old fun fact and return that instead.

Second, a few structural notes. If we do above and return early on empty userFunFact, then we don't need some of the if statements checking that anymore. Otherwise, it would be good to combine the two if-statements that have the same condition (no need for separate blocks).

Let me know if any of this is not clear, and I'm happy to explain more.

Reply:
Yes, that was the logic I intended for! I was trying to think of a way to make it easier and I guess that didn't cross my mind. Thank you for pointing that out! Please let me know if I missed something after I restructured it